BuildingBlocks/src/BuildingBlocks.Application/MediatR/QuotaEnforcerBehavior.cs

```diff
@@ -0,0 +1,51 @@
+﻿using Enmeshed.BuildingBlocks.Application.Attributes;
+using Enmeshed.BuildingBlocks.Domain.StronglyTypedIds;
+using Enmeshed.Common.Infrastructure.Persistence.Repository;
+using MediatR;
+using Enmeshed.BuildingBlocks.Application.Abstractions.Exceptions;
+using Enmeshed.BuildingBlocks.Application.Abstractions.Infrastructure.UserContext;
+
+namespace Enmeshed.BuildingBlocks.Application.MediatR;
+public class QuotaEnforcerBehavior<TRequest, TResponse> : IPipelineBehavior<TRequest, TResponse>
+    where TRequest : IRequest<TResponse>
+{
+    private readonly IMetricStatusesRepository _metricStatusesRepository;
+    private readonly IUserContext _userContext;
+
+    public QuotaEnforcerBehavior(IMetricStatusesRepository metricStatusesRepositories, IUserContext userContext)
+    {
+        _metricStatusesRepository = metricStatusesRepositories;
+        _userContext = userContext;
+    }
+
+    public async Task<TResponse> Handle(TRequest request, RequestHandlerDelegate<TResponse> next, CancellationToken cancellationToken)
+    {
+        // Using reflection.
+        var attrs = request.GetType().CustomAttributes;  // Reflection.
+
+        // Displaying output.
+        foreach (var attr in attrs)
+        {
+            if (attr.AttributeType == typeof(ApplyQuotasForMetricsAttribute))
+            {
+                var args = attr.ConstructorArguments.Select(it => new MetricKey(it.Value as string)).ToList();
+                var statuses = await _metricStatusesRepository.GetMetricStatuses(_userContext.GetAddress(), args);
+                var expiredStatuses = statuses.Where(it=> it.IsExhaustedUntil < DateTime.Now).ToList();
```

Comment:
Use `SystemTime` instead of `DateTime` please. `SystemTime` is a wrapper that allows setting the current time in order to make code that uses time testable.